Remove enumerating the list when not required

rahulpnath opened the pull request:
The cast to Array evaluates the whole list even in the case where we apply a Where condition in the IsInheritableType scenario. Don't think there is a need to break when empty list as that will happen automatically.

JimBobSquarePants commented:
Thanks @rahulpnath 

I don't see the benefit of this? By removing the break we are now always using reflection to check the type rather than breaking as soon as possible and avoiding that overhead.

rahulpnath committed:
Remove enumerating the list when not required

The cast to Array evaluates the whole list even in the case where we apply a Where condition in the IsInheritableType scenario. Don't think there is a need to break when empty list as that will happen automatically.

rahulpnath commented:
@JimBobSquarePants Ah ok. Do you see value with the break in there? Added in the break back in. Do you see value in not enumerating to an Array upfront?

JimBobSquarePants commented:
@rahulpnath Yeah that looks better. We're not really performing multiple enumeration since `Any()` jumps out immediately and we are no longer allocating a new array.

I think we can actually use the following in the first branch reassigning the IEnumerable rather than allocating a new instance now.

```
contentList = contentList.Where(c => c.DocumentTypeAlias.InvariantEquals(returnType.Name));
```

rahulpnath committed:
Removed ToArray call

rahulpnath commented:
@JimBobSquarePants Makes sense. Updated

JimBobSquarePants commented:
Perfect! Thanks! 👍
